File: src/test/java/com/raychenon/leetcode/MoveZeroesTest.java
package com.raychenon.leetcode;

import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.assertArrayEquals;

/**
 * User: raychenon
 * Date: 17/3/2020
 * https://leetcode.com/problems/move-zeroes/
 */
public class MoveZeroesTest {

    private static MoveZeroes instance;

    @BeforeClass
    public static void setUp() {
        instance = new MoveZeroes();
    }

    @Test
    public void moveZeroesTest() {
        int[] input = {0, 1, 0, 3, 12};
        int[] expected = {1, 3, 12, 0, 0};
        testWithVariousSolutions(expected, input);
    }

    @Test
    public void moveZeroesTest2() {
        int[] input = {0, 1, 0, 3, 4, 5, 6, 0, 6, 0, 0, 12};
        int[] expected = {1, 3, 4, 5, 6, 6, 12, 0, 0, 0, 0, 0};
        testWithVariousSolutions(expected, input);
    }


    @Test
    public void allZeroesBehindTest() {
        int[] input = {1, 2, 5, 3, 4, 0, 0, 0};
        int[] expected = {1, 2, 5, 3, 4, 0, 0, 0};
        testWithVariousSolutions(expected, input);
    }

    @Test
    public void allZeroesInFrontTest() {
        int[] input = {0, 0, 0, 1, 2, 5, 3, 4};
        int[] expected = {1, 2, 5, 3, 4, 0, 0, 0};
        testWithVariousSolutions(expected, input);
    }


    @Test
    public void OneZeroesInMiddleTest() {
        int[] input = {1, 2, 5, 0, 3, 22, 4};
        int[] expected = {1, 2, 5, 3, 22, 4, 0};
        testWithVariousSolutions(expected, input);
    }


    void testWithVariousSolutions(int[] expected, int[] input) {
        int[] input1 = input.clone();
        instance.moveZeroes(input1);
        assertArrayEquals(expected, input1);

        int[] input2 = input.clone();
        instance.moveZeroesLeetcodeSolution3(input2);
        assertArrayEquals(expected, input2);
    }
}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/MoveZeroesTest.kt
package com.raychenon.kotlin.leetcode

import org.junit.Assert
import org.junit.Test

/**
 * User: raychenon
 * Date: 19/3/2020
 * https://leetcode.com/problems/move-zeroes
 */
class MoveZeroesTest {

    @Test
    fun moveZeroesTest() {
        val input = intArrayOf(0, 1, 0, 3, 12)
        testWithVariousSolutions(intArrayOf(1, 3, 12, 0, 0), input)
    }

    @Test
    fun moveZeroesTest2() {
        val input = intArrayOf(0, 1, 0, 3, 5, 5, 0, 11, 12)
        testWithVariousSolutions(intArrayOf(1, 3, 5, 5, 11, 12, 0, 0, 0), input)
    }

    @Test
    fun allZeroesBehindTest() {
        val input = intArrayOf(1, 2, 5, 3, 4, 0, 0, 0)
        testWithVariousSolutions(intArrayOf(1, 2, 5, 3, 4, 0, 0, 0), input)
    }

    @Test
    fun allZeroesInFrontTest() {
        val input = intArrayOf(0, 0, 0, 1, 2, 5, 3, 4)
        testWithVariousSolutions(intArrayOf(1, 2, 5, 3, 4, 0, 0, 0), input)
    }

    @Test
    fun OneZeroesInMiddleTest() {
        val input = intArrayOf(1, 2, 5, 0, 3, 22, 4)
        testWithVariousSolutions(intArrayOf(1, 2, 5, 3, 22, 4, 0), input)
    }

    fun testWithVariousSolutions( expected: IntArray,input: IntArray): Unit{
        val input1 = input.copyOf()
        MoveZeroes.moveZeroes(input1)
        Assert.assertArrayEquals(expected, input1)

        val input2 = input.copyOf()
        MoveZeroes.moveZeroesLeetcodeSolution3(input2)
        Assert.assertArrayEquals(expected, input2)
    }
}